include/pdn.h

```diff
@@ -0,0 +1,149 @@
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ *
+ * @brief Interface for PDN connection controlling.
+ *        No data can be exchanged over a PDN interface.
+ *        The interface is used to manage connections to specific APNs.
+ */
+
+/**
+ * @file pdn.h
+ *
+ * @brief Public APIs for the PDN library.
+ */
+
+#ifndef PDN_H_
+#define PDN_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/** @brief State of the PDN
+ */
+enum pdn_state { PDN_DOWN = 0, PDN_UP = 1 };
+
+/** @brief Notify event types
+ */
+enum pdn_event {
+	PDN_EVENT_NA = 0,
+	/* TODO: TBD */
+};
+
+/**
+ * @typedefs pdn_notif_handler_t
+ *
+ * This notification handler let's application to receive notification
+ * about incoming PDN events. Notifications will be dispatched to handler
+ * registered using the @ref pdn_register_handler() function. Handlers
+ * can be de-registered using the @ref pdn_deregister_handler() function.
+ *
+ * @param[in] handle    Pointer to pdn handle.
+ * @param[in] event     Event id @ref enum pdn_event.
+ */
+typedef void (*pdn_notif_handler_t)(void *handle, const enum pdn_event event);
+
+/**
+ * @brief Routine to open a PDN channel.
+ *
+ * @return  New handle if the procedure succeeds else NULL.
+ *          errno indicates the reason for failure.
+ */
+void *pdn_open(void);
+
+/**
+ * @brief Routine to close a PDN socket.
+ *
+ * @param[in] handle    Handle returned by @ref pdn_open().
+ *
+ * @return  0 if the procedure succeeds, else, -1.
+ *          errno indicates the reason for failure.
+ */
+int pdn_close(void **handle);
+
+/**
+ * @brief Function to connect a PDN socket to a remote address.
+ *
+ * @param[in] handle           Handle returned by @ref pdn_open().
+ * @param[in] p_remote_addr    Pointer to a structure containing the remote
+ *                             address to connect to.
+ * @param[in] addrlen          Length of the destination address
+ *                             @p p_remote_addr
+ *
+ * @return  0 if the procedure succeeds, else, -1.
+ *          errno indicates the reason for failure.
+ */
+int pdn_connect(void *handle, const void *p_remote_addr, unsigned int addrlen);
+
+/**
+ * @brief Routine to set options (configurations) on a socket.
+ *
+ * @param[in] handle      Handle returned by @ref pdn_open().
+ * @param[in] level       Level, or group to which the option belongs.
+ * @param[in] optname     Name of the socket option.
+ * @param[in] p_optval    Value, passed by reference, to be stored for
+ *                        this option.
+ * @param[in] optlen      Size of @p p_optval
+ *
+ * @return  0 if the procedure succeeds, else, -1.
+ *          errno indicates the reason for failure.
+ */
+int pdn_setopt(void *handle, unsigned int level, int optname,
+	       void const *p_optval, unsigned int optlen);
+
+/**
+ * @brief Routine to get options (configurations) on a socket.
+ *
+ * @param[in]     handle     Handle returned by @ref pdn_open().
```

Comment:
void* handle is not very convenient way, imo. I have been told that I should avoid typedefs as well but can it be used here to force compiler warning/error in case of wrong type is give as a handle?
```
/* Handle to be used to control PDN related actions */
typedef void* pdn_handle;
```

Reply:
I'm afraid covering `void *` with a custom type won't make the compiler to produce any extra warnings. For the compiler, it's still `void *` so it will accept any pointer.

Reply:
@peho-nordic Only inputs that are _pointers_ should be `const` when appropriate.
It doesn't make sense to have other types of params declared `const` in a header, e.g. `foo(const int n)`: neither the caller or compiler will care about that `const`.

Reply:
yeah, that is modified now and handle is integer value.